### ISSUE TITLE
[BUG][JS] Generated JS SDK result in the following error when called : end() was called twice. This is not supported in superagent

### ISSUE THREAD

**Issue by Thytu** [OpenAPITools/openapi-generator#12073]
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [X] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [X] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [X] Have you searched for related issues/PRs?
- [X] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

I use FastAPI to generate my routes, and the following command to produce the sdk:
```sh
docker run --network=host --rm \
        -v $(pwd)/clients:/tmp \
        openapitools/openapi-generator-cli:v4.0.0 generate \
        -i localhost:8080/openapi.json \
        -o /tmp/{language} \
        -D modelDocs=false \
        -D apiDocs=false \
        -D apiTests=false \
        -D modelTests=false \
        -D npmVersion=3.5.2 \
        -D supportsES6=true \
        -g javascript
```

When producing the following code it result into this error message : `.end() was called twice. This is not supported in superagent`

Executed code:
```js
useEffect(() => {

    const api =  new MyApi()

    api.apiClient.basePath = 'http://localhost:8080'

    api.produceSomethigGet().then(res => {
      console.log(res)
    })
  }, [])
```

##### openapi-generator version

V4.0.0

##### OpenAPI declaration file content or url

https://gist.github.com/Thytu/517d892fadd685c921e7b0d780b1cd21

##### Steps to reproduce

Generate SDK:
```sh
docker run --network=host --rm \
        -v $(pwd)/clients:/tmp \
        openapitools/openapi-generator-cli:v4.0.0 generate \
        -i localhost:8080/openapi.json \
        -o /tmp/{language} \
        -D modelDocs=false \
        -D apiDocs=false \
        -D apiTests=false \
        -D modelTests=false \
        -D npmVersion=3.5.2 \
        -D supportsES6=true \
        -g javascript
```

Call SDK:
```js
useEffect(() => {

    const api =  new MyApi()

    api.apiClient.basePath = 'http://localhost:8080'

    api.produceSomethigGet().then(res => {
      console.log(res)
    })
  }, [])
```

##### Suggest a fix

I saw that if I change a bit the client code it does word:

Unchanged `callApi` function:
```js
callApi(path, httpMethod, pathParams,
        queryParams, headerParams, formParams, bodyParam, authNames, contentTypes, accepts,
        returnType, apiBasePath, callback) {

        var url = this.buildUrl(path, pathParams, apiBasePath);
        var request = superagent(httpMethod, url);

        if (this.plugins !== null) {
            for (var index in this.plugins) {
                if (this.plugins.hasOwnProperty(index)) {
                    request.use(this.plugins[index])
                }
            }
        }

        // apply authentications
        this.applyAuthToRequest(request, authNames);

        // set query parameters
        if (httpMethod.toUpperCase() === 'GET' && this.cache === false) {
            queryParams['_'] = new Date().getTime();
        }

        request.query(this.normalizeParams(queryParams));

        // set header parameters
        request.set(this.defaultHeaders).set(this.normalizeParams(headerParams));

        // set requestAgent if it is set by user
        if (this.requestAgent) {
          request.agent(this.requestAgent);
        }

        // set request timeout
        request.timeout(this.timeout);

        var contentType = this.jsonPreferredMime(contentTypes);
        if (contentType) {
            // Issue with superagent and multipart/form-data (https://github.com/visionmedia/superagent/issues/746)
            if(contentType != 'multipart/form-data') {
                request.type(contentType);
            }
        } else if (!request.header['Content-Type']) {
            request.type('application/json');
        }

        if (contentType === 'application/x-www-form-urlencoded') {
            request.send(querystring.stringify(this.normalizeParams(formParams)));
        } else if (contentType == 'multipart/form-data') {
            var _formParams = this.normalizeParams(formParams);
            for (var key in _formParams) {
                if (_formParams.hasOwnProperty(key)) {
                    if (this.isFileParam(_formParams[key])) {
                        // file field
                        request.attach(key, _formParams[key]);
                    } else {
                        request.field(key, _formParams[key]);
                    }
                }
            }
        } else if (bodyParam !== null && bodyParam !== undefined) {
            request.send(bodyParam);
        }

        var accept = this.jsonPreferredMime(accepts);
        if (accept) {
            request.accept(accept);
        }

        if (returnType === 'Blob') {
          request.responseType('blob');
        } else if (returnType === 'String') {
          request.responseType('string');
        }

        // Attach previously saved cookies, if enabled
        if (this.enableCookies){
            if (typeof window === 'undefined') {
                this.agent._attachCookies(request);
            }
            else {
                request.withCredentials();
            }
        }

        request.end((error, response) => {
            if (callback) {
                var data = null;
                if (!error) {
                    try {
                        data = this.deserialize(response, returnType);
                        if (this.enableCookies && typeof window === 'undefined'){
                            this.agent._saveCookies(response);
                        }
                    } catch (err) {
                        error = err;
                    }
                }

                callback(error, data, response);
            }
        });

        return request;
    }
```

Changed `callApi` function:
```js
   callApi(path, httpMethod, pathParams,
        queryParams, headerParams, formParams, bodyParam, authNames, contentTypes, accepts,
        returnType, apiBasePath, callback) {

        var url = this.buildUrl(path, pathParams, apiBasePath);
        var request = superagent(httpMethod, url);

        if (this.plugins !== null) {
            for (var index in this.plugins) {
                if (this.plugins.hasOwnProperty(index)) {
                    request.use(this.plugins[index])
                }
            }
        }

        // apply authentications
        this.applyAuthToRequest(request, authNames);

        // set query parameters
        if (httpMethod.toUpperCase() === 'GET' && this.cache === false) {
            queryParams['_'] = new Date().getTime();
        }

        request.query(this.normalizeParams(queryParams));

        // set header parameters
        request.set(this.defaultHeaders).set(this.normalizeParams(headerParams));

        // set requestAgent if it is set by user
        if (this.requestAgent) {
          request.agent(this.requestAgent);
        }

        // set request timeout
        request.timeout(this.timeout);

        var contentType = this.jsonPreferredMime(contentTypes);
        if (contentType) {
            // Issue with superagent and multipart/form-data (https://github.com/visionmedia/superagent/issues/746)
            if(contentType != 'multipart/form-data') {
                request.type(contentType);
            }
        } else if (!request.header['Content-Type']) {
            request.type('application/json');
        }

        if (contentType === 'application/x-www-form-urlencoded') {
            request.send(querystring.stringify(this.normalizeParams(formParams)));
        } else if (contentType == 'multipart/form-data') {
            var _formParams = this.normalizeParams(formParams);
            for (var key in _formParams) {
                if (_formParams.hasOwnProperty(key)) {
                    if (this.isFileParam(_formParams[key])) {
                        // file field
                        request.attach(key, _formParams[key]);
                    } else {
                        request.field(key, _formParams[key]);
                    }
                }
            }
        } else if (bodyParam !== null && bodyParam !== undefined) {
            request.send(bodyParam);
        }

        var accept = this.jsonPreferredMime(accepts);
        if (accept) {
            request.accept(accept);
        }

        if (returnType === 'Blob') {
          request.responseType('blob');
        } else if (returnType === 'String') {
          request.responseType('string');
        }

        // Attach previously saved cookies, if enabled
        if (this.enableCookies){
            if (typeof window === 'undefined') {
                this.agent._attachCookies(request);
            }
            else {
                request.withCredentials();
            }
        }

        // request.end((error, response) => {
        //     if (callback) {
        //         var data = null;
        //         if (!error) {
        //             try {
        //                 data = this.deserialize(response, returnType);
        //                 if (this.enableCookies && typeof window === 'undefined'){
        //                     this.agent._saveCookies(response);
        //                 }
        //             } catch (err) {
        //                 error = err;
        //             }
        //         }

        //         callback(error, data, response);
        //     }
        // });

        return request;
    }
```

**Comment by ccaspanello** [comment]
I am also experiencing the same issue.

**Comment by alex-tw-lam** [comment]
The generated code has two ways to handle making a network request:
- The Callback Way (Old Style): You give the code a function (a callback) to run after the request finishes. This is the expected way when the code is generated with the default setting (usePromises=false).
- The Promise Way (New Style): The code gives you back a special object. You can attach .then() to this object to say what should happen after the request finishes.

Problem:
- When the code is generated with the default setting (usePromises=false), it's supposed to use the Callback Way. 
- However, the generated code has a flaw:
  - It always starts the request internally using the Callback Way (by calling request.end()).
  - At the same time, it also gives you back the special object that looks like it can be used with the Promise Way (.then()).
- Basically, when the client is generated with usePromises=false, it incorrectly appears to handle Promises. 


So, if you don't provide a callback and instead try to use .then() (the Promise Way), the underlying network library (superagent) gets confused. 
- It sees the request being started twice for the same operation: once internally via the Callback Way, and once by your .then() call. 
- This confusion causes it to throw the "request sent twice" error.

To generate an SDK that uses Promises:
```
podman run --rm -v ${PWD}:/local openapitools/openapi-generator-cli generate \
    -i <your_openapi_spec.yaml>  \
    -g javascript \
    --additional-properties=usePromises=true \
    -o /local/<your_api_client_dir>
```
